### ISSUE TITLE
Change order and setup dependencies in passthrough

### ISSUE THREAD

**Issue by tokejepsen** [pyblish/pyblish-deadline#27]
**Goal**

Get ```Deadline Submission``` to process instances, instead of context, so its easier to see how far the submission process is.

**Implemetation**

Change the order of the instances in the context in what is now ```Deadline Passthrough```. This will ensure the processing of the instances are in the correct order, so setting up dependencies in ```Deadline Submission``` is possible without processing the context.

@mottosso guessing you can change the order of the instance in context, similar to how you might do it in a collector?

**Comment by mottosso** [comment]
> @mottosso guessing you can change the order of the instance in context, similar to how you might do it in a collector?

Yes, absolutely. Or where are you thinking you would change the order here? In an extractor..? You can change them anytime a plug-in starts processing I think. I think it fetches the order once a plug-in starts running, and then remembers it until the next plug-in. But haven't tried so would have to investigate to be sure.

**Comment by tokejepsen** [comment]
> Or where are you thinking you would change the order here? In an extractor..?

Yeah, I was thinking that I change the order in a plugin prior to the submission plugin.

I guess its a bit of a technicality atm with the latest process indicators, but down the line it would look more like a progress bar and so its good for users to know how far the processing is.

**Comment by mottosso** [comment]
Ah. Yeah, I think it shouldn't really make a difference.

A neat test would be to change the order at one point, print it out, and then change it back in some other plug-in down the line. Just to see how much control you could get out of it. Perhaps there's a potential benefit of having an order tailored per plug-in to some extent; maybe in terms of performance, maybe in terms or parallelism once we get to that point.

I think no matter the order, all instances will always get processed, so at the end of a publish, it shouldn't matter how or how often they get rearranged during that time.

Interesting idea.

**Comment by mottosso** [comment]
Actually now that I think of it, the order will likely be the visual order from QML. The idea was to enable re-arranging them visually.

It's an easy switch (in the publish() and validate() functions), but we'll have to take that into account if changing that behaviour. Which is better, or whether it's worth having a toggle for it.

**Comment by tokejepsen** [comment]
The current version of pyblish doesn't allow for rearranging the order of the instances when processing, only when collecting.

```python
import pyblish.api
import pyblish_integration


class CollectorA(pyblish.api.Collector):

    def process(self, context):
        for x in range(0, 4):
            instance = context.create_instance(str(x), family="A")


class CollectorB(pyblish.api.Collector):

    def process(self, context):
        context.reverse()


class Validator(pyblish.api.Validator):

    def process(self, context):
        context.reverse()
            

class Extractor(pyblish.api.Extractor):
    families = ["A"]
    
    def process(self, instance):
        self.log.info(instance)


pyblish.api.register_plugin(CollectorA)
pyblish.api.register_plugin(CollectorB)
pyblish.api.register_plugin(Validator)
pyblish.api.register_plugin(Extractor)
pyblish_integration.show()
```

In the above example you can comment out ```CollectorB``` and the order being printed in ```Extractor``` won't change.

So you are suggesting that we need to modify the pyblish-base for this to happen? Do you think its a good feature to add? The only use-case atm is this.

**Comment by tokejepsen** [comment]
Turns out this actually works, it was just wrong of me to look at the GUI. This code shows that you can reorder the context for other plugins;

```python
import pyblish.api
import pyblish.util


pyblish.api.deregister_all_plugins()

class CollectorA(pyblish.api.Collector):

    def process(self, context):
        for x in range(0, 4):
            instance = context.create_instance(str(x), family="A")


class ValidatorA(pyblish.api.Validator):

    def process(self, context):
        print context


class ValidatorB(pyblish.api.Validator):

    def process(self, context):
        context.reverse()


class ValidatorC(pyblish.api.Validator):

    def process(self, context):
        print context


pyblish.api.register_plugin(CollectorA)
pyblish.api.register_plugin(ValidatorA)
pyblish.api.register_plugin(ValidatorB)
pyblish.api.register_plugin(ValidatorC)
pyblish.util.publish()
```

**Comment by tokejepsen** [comment]
The problem is still with the GUI, that remembers the order from collection. With the recent PR to pyblish-qml, we could trigger the [refresh method](https://github.com/pyblish/pyblish-qml/blob/95444f93a69704cee764175769620d772355899a/pyblish_qml/control.py#L528) during processing to update the context order in the GUI.
The question is how/when?

The simple solution I'd imagine would be doing a refresh after each plugin. But it will be a major performance hit (i think), and very unnecessary.

The complex solution would be to introduce a flag on a plugin, that tells the GUI to refresh. This would target the refresh method processing to very specific plugins. The flag could look something like this;

```python
class ValidatorB(pyblish.api.Validator):

    gui_refresh = True

    def process(self, context):
        context.reverse()
```

@mottosso what do you think?

**Comment by mottosso** [comment]
Refreshing per process does sound like a performance hit, but it's worth testing to see if it actually is.

If so, it might be worth looking at a custom field being sent to QML along with each [result](http://api.pyblish.com/pages/result.html), with just the names of instances along with their order, such as in a simple list of strings.

If you are up for testing a refresh per process, evaluate the performance, we could continue having a chat about how to make it faster.

**Comment by tokejepsen** [comment]
Have you got any good ideas about a performance test?

**Comment by tokejepsen** [comment]
I'm definitely seeing some performance differences, but wanted to document how much.

**Comment by mottosso** [comment]
`timeit` is a pretty good tool, you could setup a test that performs the operation, and run it under the two different circumstances.

- https://docs.python.org/2/library/timeit.html

**Comment by tokejepsen** [comment]
I ended up running the ```run_testsuite.py``` a couple of times and averaging.

~2.13s - with refresh
~2.02s - without refresh

Doesn't sound like a lot, but it noticeable in more complex publishes.

What do you think? Optimize or explicit refreshing?

**Comment by mottosso** [comment]
Yeah, it will be more complex with more instances and data in each instance..

Let's have a think about how to add this feature to the GUI properly. Sounds like a forum thread is in order.
